### PR TITLE
P751sike

### PR DESCRIPTION
**Pull request by amirjalali65**
Incorporated the comments.
sha3 package and cSHAKE gist are now included in the repo.

### CODE REVIEW

File: sike.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+// the key based on NIST specifications. 
+//
+//Author: Amir Jalali  				ajalali2016@fau.edu
+//Date: Feb 2018
+
+package p751sidh
+
+import (
+	"io"
+	"golang.org/x/crypto/sha3"
```
Comment (grittygrease) [inline sike.go]: Shouldn't this be:
"github.com/cloudflare/p751sidh/sha3" ?

File: sike.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	Scalar [MessageSize]byte
+}
+
+type SIKESharedSecret struct {
+	Scalar [SharedSecretSize]byte
+}
+
+// SIKE keypair generation generates SIKE secret-key and public-key.
+// The secret-key contains a random message + secret-key + public-key.   
+func GenerateKeyPair(rand io.Reader) (publicKey *SIDHPublicKeyBob, SIKEsecretKey *SIKESecretKey, err error) {
```
Comment (grittygrease) [inline sike.go]: If this function name is "GenerateKeyPair" but the package is p751sidh, then it shares a namespace with GenerateAliceKeyPair and other SIDH-specific functions. I recommend moving this file into its own folder so that it can be an independent module.

File: sike.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	
+	publicKey = new(SIDHPublicKeyBob)
+	SIKEsecretKey = new(SIKESecretKey)
```
Comment (grittygrease) [inline sike.go]: *sikeSecretKey, per https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: sike.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+import (
+	"io"
+	"golang.org/x/crypto/sha3"
+	"bytes"
+)
+
+const (
+	// The message size, in bytes.
+	MessageSize = 32
```
Comment (grittygrease) [inline sike.go]: This code is specialized to p751 (as is the rest of the code in the p751sidh package). Since SIKE is a general construction that takes a SIDH implementation and builds upon it. Maybe a refactor of this repo so that sike takes a general SIDH struct would make sense. For example, https://golang.org/pkg/crypto/elliptic/#Curve has a Curve object and that is used in https://golang.org/pkg/crypto/ecdsa/#GenerateKey. This could make the package more generally useful once there are other SIDH implementations in Go other than p751sidh.

**Comment by Bren2010** [comment]
Where on earth did sha3/ come from? Why wouldn't we just import it from there?

Why did you open a second PR with the same name?

File: sidh/sidh.go [hunk 5]
```diff
@@ -12,7 +12,7 @@
 //
 // If you feel that SIDH may be appropriate for you, consult your
 // cryptographer.
-package p751sidh
```
Comment (grittygrease) [inline sidh/sidh.go]: This package is renamed sidh, but there are many p751-specific variables and even an include of p751toolbox. If you're going to make this generic, can you remove the p751-specific code and variables? If you're going to make a generic package, it should have only the sidh-specific types and interfaces:
 type SIDHPublicKeyBob struct {}
 func (pubKey *SIDHPublicKeyBob) FromBytes(input []byte)
 func GenerateAliceKeypair(rand io.Reader) (publicKey *SIDHPublicKeyAlice, secretKey *SIDHSecretKeyAlice, err error)
 etc.

Actual implementations should be part of a curve-specific package (like sidh/p751sidh) that instantiates a function that fulfils the interfaces for SIDHPublicKeyAlice, and other structures. Unless these functions can be written generically for a specific SIDH implementation, they probably don't belong.

File: sha3/cshake.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+type CShakeHash interface {
+	io.Writer
+	io.Reader
+	Clone() ShakeHash
+	Reset()
+}
+
+// NewcShake128 creates a new custom cSHAKE128 variable-output-length CShakeHash.
```
Comment (grittygrease) [inline sha3/cshake.go]: Can you split this cshake implementation into a separate commit?

File: sidh/sidh_test.go [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package p751sidh
+package sidh
```
Comment (grittygrease) [inline sidh/sidh_test.go]: It's actually fine to have the p751toolbox used in the testing of the sidh package, even if the sidh package is stripped down to just the function definitions.

File: sike/sike.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+		return nil, nil, err
+	}
+
+	// Append publicKey to message and hash it
+	publicKey.PublicKey.ToBytes(tmp[MessageSize:])
+	CShakeSum256(ephemeralSk.Scalar[:], tmp[:CiphertextSize], []byte(G))
+
+	// Perform mod oA
+	ephemeralSk.Scalar[47] = 0
+	ephemeralSk.Scalar[46] &= 15 // clear high bits, so scalar < 2^372
```
Comment (grittygrease) [inline sike/sike.go]: Is there any way to make this generic to any SIDH implementation?

File: sidh/sidh.go (hunk 5 above)
Comment (amirjalali65) [reply sidh/sidh.go]: It is possible to include all the p751-specific variables inside consts.go file or create another file which contains these values and sidh package simply retrieve these values from p751toolbox package. I will include it in the next commit.

File: sha3/cshake.go (hunk 6 above)
Comment (amirjalali65) [reply sha3/cshake.go]: Sure

File: sidh/sidh_test.go (hunk 7 above)
Comment (amirjalali65) [reply sidh/sidh_test.go]: Noted. I will incorporate it

File: sike/sike.go (hunk 8 above)
Comment (amirjalali65) [reply sike/sike.go]: I think it is possible to define constant values such as AliceMask and BobMask in consts.go file and replace these numbers with these variables. I will include it in the next commit

File: sidh/sidhp751/sidhp751.go [hunk 9]
```diff
@@ -59,68 +94,44 @@ var bobIsogenyStrategy = [maxBob]int{0, 1, 1, 2, 2, 2, 3, 3, 4, 4, 4, 5, 5, 5, 6
...
 
-// Bob's public key.
-type SIDHPublicKeyBob struct {
-	affine_xP   ExtensionFieldElement
-	affine_xQ   ExtensionFieldElement
-	affine_xQmP ExtensionFieldElement
-}
-
 // Read a public key from a byte slice.  The input must be at least 564 bytes long.
 func (pubKey *SIDHPublicKeyBob) FromBytes(input []byte) {
```
Comment (grittygrease) [inline sidh/sidhp751/sidhp751.go]: This looks like it could be a generic function in the sidh package.

File: sidh/sidhp751/sidhp751.go [hunk 10]
```diff
@@ -136,9 +147,9 @@ func GenerateAliceKeypair(rand io.Reader) (publicKey *SIDHPublicKeyAlice, secret
 	}
 
 	// Bit-twiddle to ensure scalar is in 2*[0,2^371):
-	secretKey.Scalar[47] = 0
-	secretKey.Scalar[46] &= 15 // clear high bits, so scalar < 2^372
-	secretKey.Scalar[0] &= 254 // clear low bit, so scalar is even
+	secretKey.Scalar[SecretKeySize-1] = MaskAliceByte1
+	secretKey.Scalar[SecretKeySize-2] &= MaskAliceByte2 // clear high bits, so scalar < 2^372
+	secretKey.Scalar[0] &= MaskAliceByte3               // clear low bit, so scalar is even
```
Comment (grittygrease) [inline sidh/sidhp751/sidhp751.go]: Question: is this masking generic or specific to p751?

File: sidh/sidhp751_interface.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
+// sidhp751_interface implements all the required interface APIs for performing
+// SIDH key-exchange over p751 parameter. It imports the sidhp751 implementation
+// and converts all the underlying data structures to generic interface structures.
+// Author: Amir Jalali		ajalali2016@fau.edu
+package sidh
```
Comment (grittygrease) [inline sidh/sidhp751_interface.go]: This code should probably be in the sidhp751 package.

File: sike/sikep751/sikep751.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+// SIKESharedSecret contains SIDHSecretKeyBob, SIDHPublicKeyBob, and Scalar
+type SIKESharedSecret struct {
+	Scalar [SharedSecretSize]byte
+}
+
+// GenerateKeyPair generation generates SIKE secret-key and public-key.
+// The secret-key contains a random message + secret-key + public-key.
+func GenerateKeyPair(rand io.Reader) (publicKey *SIKEPublicKey, sikeSecretKey *SIKESecretKey, err error) {
```
Comment (grittygrease) [inline sike/sikep751/sikep751.go]: This generic SIKE code should probably live in the sike package.

File: sike/sikep751_interface.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+// Author: Amir Jalali		ajalali2016@fau.edu
+package sike
+
+import (
+	"io"
+)
+
+import . "github.com/cloudflare/p751sidh/sike/sikep751"
+
+type p751SIKE struct {
```
Comment (grittygrease) [inline sike/sikep751_interface.go]: This should be in the sikep751 package, since it's specific to p751.

File: sidh/sidhp751_interface.go (hunk 11 above)
Comment (grittygrease) [reply sidh/sidhp751_interface.go]: Actually, the functions GenerateAliceKeypair and such are generic so should be in this package, but the `func (e *p751SIDH)` functions are generic and don't need to be written because they can be defined generically on the SIDH struct in sidh.go.

File: sidh/sidhp751/sidhp751.go (hunk 9 above)
Comment (amirjalali65) [reply sidh/sidhp751/sidhp751.go]: It is one of the things that I wasn't sure about. Since in any other DH key-exchange package this kind of functionality is not available, I thought it would make more sense to leave it in the lower implementation. But I will add it to the sidh package since we need to use it inside sike implementation for instance.

File: sidh/sidhp751/sidhp751.go (hunk 10 above)
Comment (amirjalali65) [reply sidh/sidhp751/sidhp751.go]: It's specific to p751 and Alice and Bob torsion subgroup order

File: newsidh/sidh_test.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
+package sidh
+
+import (
+	"bytes"
```
Comment (henrydcase) [inline newsidh/sidh_test.go]: test

File: sike/sike.go [hunk 15]
```diff
@@ -0,0 +1,33 @@
+// Package sike contains all the required interfaces for running an instance of
+// Supersingular Isogeny Key Encapsulation (SIKE) mechanism between two parties.
+// This package provide all the required APIs and generic data-types for SIKE
```
Comment (henrydcase) [inline sike/sike.go]: Can you rephrase it to that:
"""
This package contains implementation of Supersingular Isogeny Key Encapsulation (SIKE) protocol. SIKE is built on top of SIDH and implements KEM encapsulation and KEM decapsulation. 
Both, data types and API. provided by this package are generic and can be instantiated with specific parameters (see sikep751.go).
Author: Amir Jalali		ajalali2016@fau.edu
"""

File: sike/sike_test.go [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+	testSIKE(NewP751SIKE(), t)
+}
+
+func BenchmarkP751SIKE(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		testSIKE(NewP751SIKE(), b)
+	}
+}
+
+func testSIKE(sikep751 SIKE, t testing.TB) {
```
Comment (henrydcase) [inline sike/sike_test.go]: I would really like to see at least one KAT here (decapsulation). I think NIST submission contains some KATs (key, shared secret and ct).

File: sha3/cshake.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	cShakePad = append(cShakePad, []byte{1, 0}...) // left_encode(0)
+	cShakePad = append(cShakePad, left_encode(uint64(len(custom) * 8))...)
+	cShakePad = append(cShakePad, custom...)	
+	padding_len := cshake.rate - (len(cShakePad) % cshake.rate)
+	cShakePad = append(cShakePad, bytes.Repeat([]byte{0}, padding_len)...)
+	cshake.Write(cShakePad)
+}
+
+// Helper function for the initialization of cShake
+func left_encode(value uint64) []byte {
```
[the remark this answers is not in the document]
Comment (henrydcase) [reply sha3/cshake.go]: I think this function is a bit weird. I think it can be simplified:
```
// Helper function for the initialization of cShake
func leftEncode(value uint64) []byte {

    var n uint
    var b [9]byte

    for v:=value; v!=0; v = v>>8 {
    	n++
    }

    if n==0 {
    	n = 1
    }

    for i:=uint(1); i<=n; i++ {
    	b[i] = byte(value >> (8*(i-1)))
    }

    b[0] = byte(n)
    return b[:n+1]
}
```

File: sike/sikep751/sikep751.go [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+package sikep751
+
+import (
+	"bytes"
+	"io"
+)
+
+import . "github.com/cloudflare/p751sidh/sha3"
+
+import . "github.com/cloudflare/p751sidh/sidh/sidhp751"
```
Comment (henrydcase) [inline sike/sikep751/sikep751.go]: It won't work as "sidh" directory doesn't exist. In fact the import should stay as it is, but directory structure should be reorganized
This would be better if imports are in one block:
```
import (
	"bytes"
	"io"

	. "github.com/cloudflare/p751sidh/sha3"
	. "github.com/cloudflare/p751sidh/sidh/sidhp751"
)
```

File: sike/sikep751_interface.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
+// sikep751_interface implements all the required interface APIs for performing
```
Comment (henrydcase) [inline sike/sikep751_interface.go]: Why this file is not in sikep751? It is specific top751...

File: sike/sikep751_interface.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+type p751SIKE struct {
+	SIKE
+}
+
+func NewP751SIKE() SIKE {
+	return &p751SIKE{}
+}
+
+func (e *p751SIKE) GenerateKeyPair(rand io.Reader) (publicKey PublicKey, secretKey SecretKey, err error) {
```
Comment (henrydcase) [inline sike/sikep751_interface.go]: I don't understand why this function has more than 1 line
```
func (e *p751SIKE) GenerateKeyPair(rand io.Reader) (publicKey PublicKey, secretKey SecretKey, err error) {
	return GenerateKeyPair(rand)
}
```

File: sike/sike.go [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+type SecretKey interface {}
+
+// PublicKey type for the key-encapsulation mechanism.
+type PublicKey interface {}
+
+// Ciphertext type for the key-encapsulation mechanism.
+type Ciphertext interface {}
+
+// SIKE interface contains all the required APIs for running an instance of SIKE.
+type SIKE interface {
```
Comment (henrydcase) [inline sike/sike.go]: Improvement: it would be worth having functions to import/export keys from/to hex

File: newsidh/sidh_test.go [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+	"crypto/rand"
+	"fmt"
```
Comment (henrydcase) [inline newsidh/sidh_test.go]: I don't think it will compile if we import "fmt" which is not used anywhere

File: newsidh/sidh_test.go [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+	"testing"
+
+	. "github.com/cloudflare/p751sidh/p751toolbox"
+)
+
+func newp751SIDH() SIDH {
```
Comment (henrydcase) [inline newsidh/sidh_test.go]: I think this should go to ``sidhp751.go``

File: p751toolbox/consts.go [hunk 24]
```diff
@@ -14,3 +16,52 @@ var Affine_yPB = PrimeFieldElement{A: Fp751Element{0xd1e1471273e3736b, 0xf9301ba
 
 // The value of (a+2)/4 for the starting curve E_0 with a=0: this is 1/2
 var E0_aPlus2Over4 = PrimeFieldElement{A: Fp751Element{0x124d6, 0x0, 0x0, 0x0, 0x0, 0xb8e0000000000000, 0x9c8a2434c0aa7287, 0xa206996ca9a378a3, 0x6876280d41a41b52, 0xe903b49f175ce04f, 0xf8511860666d227, 0x4ea07cff6e7f}}
+
+const (
```
Comment (henrydcase) [inline p751toolbox/consts.go]: Everything from this point until end of the file must be removed - it is already in "sidhp751/sidhp751.go"

File: newsidh/sidh.go [hunk 25]
```diff
@@ -0,0 +1,403 @@
...
+func (e *SIDH) GenerateAliceKeypair(rand io.Reader) (publicKey *SIDHPublicKeyAlice, secretKey *SIDHSecretKeyAlice, err error) {
+	publicKey = new(SIDHPublicKeyAlice)
+	secretKey = new(SIDHSecretKeyAlice)
+
+	_, err = io.ReadFull(rand, secretKey.Scalar[:])
+	if err != nil {
+		return nil, nil, err
+	}
+	// Bit-twiddle to ensure scalar is in 2*[0,2^371):
+	secretKey.Scalar[e.secretKeySize-1] = e.maskAliceByte1
```
Comment (henrydcase) [inline newsidh/sidh.go]: I know that it's a detail but strictly speaking, secret key should be chosen uniformly at random. By applying this mask we favour some numbers over some others

File: newsidh/sidh.go [hunk 26]
```diff
@@ -0,0 +1,403 @@
...
+
+	// Perform rejection sampling to obtain a random value in [0,3^238]:
+	var ok uint32
+	for i := 0; i < e.sampleRate; i++ {
+		_, err = io.ReadFull(rand, secretKey.Scalar[:])
+		if err != nil {
+			return nil, nil, err
+		}
+		// Mask the high bits to obtain a uniform value in [0,2^378):
+		secretKey.Scalar[e.secretKeySize-1] &= e.maskBobByte
```
Comment (henrydcase) [inline newsidh/sidh.go]: similar comment as for generating secret key for "Alice"

**Comment by henrydcase** [comment]
Review done on my side.
Even though I had few comments, I think it is very interesting development. Thanks @amirjalali65

File: sike/sikep751/sikep751.go [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+
+	// Encryption
+	var tmpPk = new(SIDHPublicKeyAlice)
+	*tmpPk = ephemeralSk.PublicKey()
+	cipherText.PublicKey = tmpPk
+	jinvariant = ephemeralSk.SharedSecret(publicKey.PublicKey)
+	CShakeSum256(h[:], jinvariant[:], []byte(P))
+
+	for i := 0; i < MessageSize; i++ {
+		cipherText.Scalar[i] = tmp[i] ^ h[i]
```
Comment (henrydcase) [inline sike/sikep751/sikep751.go]: i think it should be 
"
ciphertext.Scalar[i + 596] = tmp[i] ^ h[i]
"
as in Microsoft's implementation

File: sike/sikep751/sikep751.go [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	// Generate shared secret: ss = H(m||ct)
+	cipherText.PublicKey.ToBytes(tmp[MessageSize:])
+	copy(tmp[CiphertextSize:], cipherText.Scalar[:])
```
Comment (henrydcase) [inline sike/sikep751/sikep751.go]: I think that's a bug.
1. Public key is serialized to &tmp[32] and it wrties 596 bytes
2. Then scalar is coppied to the location &tmp[596] an 32 bytes are written. Overwritting 1.

Line 123 and 124 seem are not needed, it should be:
```
copy(tmp[MessageSize:], cipherText.Scalar[:])
```

File: sike/sikep751/sikep751.go [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+// SIKE implementation parameters correspond to p751
+const (
+	// MessageSize in bytes.
+	MessageSize = 32
+	// CiphertextSize in bytes = SIDHPublicKeySize + MessageSize
+	CiphertextSize = 596
+	// SIKESecretKeySize in bytes = MessageSize + SIDHPublicKeySize + SIDHSecretKeySize
+	SIKESecretKeySize = 644
+	// G is a custom vallue for cSHAKE256
+	G = "0"
```
Comment (henrydcase) [inline sike/sikep751/sikep751.go]: As per SIKE submission to PQ NIST (see 1.4) those values should be
G = 0x00 (not 0x30)
H = 0x01 (not 0x31)
F = 0x02 (not 0x32)

**Comment by henrydcase** [comment]
I'm closing it as SIKE is already implemented.
